### PR TITLE
feat: Copy object tags [FC-0062]

### PR DESCRIPTION
**Pull request by ChrisChV**
This PR adds:

- A function `copy_tags` in `api.py` to copy tags between objects.
- A new field `is_copied` in `ObjectTag`

## Supporting information

- Github issue: https://github.com/openedx/modular-learning/issues/177
- It's used in: https://github.com/openedx/edx-platform/pull/35596
- Internal ticket: [FAL-3616](https://tasks.opencraft.com/browse/FAL-3616)

## Testing instructions

- Verify that the tests cover the new code.
- Follow the testing instructions in https://github.com/openedx/edx-platform/pull/35596

## Deadline

ASAP

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @ChrisChV!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@axim-engineering`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: openedx_tagging/core/tagging/api.py
Comment (navinkarkera) [inline openedx_tagging/core/tagging/api.py:505]: Can we rename this to something like `dest_object_has_tag`?

File: openedx_tagging/core/tagging/api.py
Comment (navinkarkera) [inline openedx_tagging/core/tagging/api.py:516]: `object_tags` to `source_tags` or `source_object_tags`?

File: tests/openedx_tagging/core/tagging/test_api.py
Comment (navinkarkera) [inline tests/openedx_tagging/core/tagging/test_api.py:941]:
```suggestion

```
Is `full_clean` required here?

**Review by navinkarkera — APPROVED** [review]
:+1: @ChrisChV Perfect :100: 
- [x] I tested this: (Copied tags, deleted tags, added same tag in source and dest and copied them.)
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

File: tests/openedx_tagging/core/tagging/test_api.py
Comment (ChrisChV) [reply tests/openedx_tagging/core/tagging/test_api.py:941]: Yes, we use `full_clean()` to validate the object tag in the test. It's used in other tests: [[1](https://github.com/openedx/openedx-learning/blob/main/tests/openedx_tagging/core/tagging/test_api.py#L430)][[2](https://github.com/openedx/openedx-learning/blob/main/tests/openedx_tagging/core/tagging/test_api.py#L449)][[3](https://github.com/openedx/openedx-learning/blob/main/tests/openedx_tagging/core/tagging/test_api.py#L479)]

File: openedx_tagging/core/tagging/api.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/6b73ec4ff4c282a007eea3fe231b9db351faf9a7

File: openedx_tagging/core/tagging/api.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/6b73ec4ff4c282a007eea3fe231b9db351faf9a7

File: openedx_tagging/core/tagging/api.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/api.py:510]: I think it's better to match using `taxonomy_id` and `tag_id` rather than `_export_id` and `_value`, assuming that we don't need to support deleted tags. See below for my recommendation to use `update_or_create`

**Review by bradenmacdonald — CHANGES_REQUESTED** [review]
Suggested some changes. Thinking ahead, we may also need an API to "convert" the tags if/when the link is broken. That is, if the user links some library into a course, then later breaks the link, the tags that were read-only would need to be changed to `is_copied=False` so that they can be edited. But we can do that in a separate PR if you prefer.

File: openedx_tagging/core/tagging/models/base.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/models/base.py:802]: I think this should not be here. If the tagging app is reused in another project other than edx-platform, a copied tag does not necessarily mean it is read-only in the UI.

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py:179]: I do it this way so that that serializer can be overridden. In my other [PR](https://github.com/openedx/edx-platform/pull/35596/files#diff-68d305b73e25deae855bf8da30edd62fcadb6cc69c61288139359ff2988ffec7R105) I override that serializer to override `get_can_delete_objecttag` and prevent the user from deleting copied tags.

File: openedx_tagging/core/tagging/rest_api/v1/views.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/rest_api/v1/views.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/744a539b9a1abfd79a698943e12468de029cbc9f

File: openedx_tagging/core/tagging/api.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/744a539b9a1abfd79a698943e12468de029cbc9f

File: openedx_tagging/core/tagging/api.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/744a539b9a1abfd79a698943e12468de029cbc9f

File: openedx_tagging/core/tagging/api.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/744a539b9a1abfd79a698943e12468de029cbc9f

File: openedx_tagging/core/tagging/api.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/api.py:517]: Did you confirm this is working? (That `_value` and `_export_id` are getting set automatically?)

**Review by bradenmacdonald — APPROVED** [review]

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (bradenmacdonald) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py]: Would be good to mention this in a comment: "# Allows consumers like edx-platform to override this"

**Comment by bradenmacdonald** [comment]
@ChrisChV I can merge this when you fix the comments and reply - let me know.

File: openedx_tagging/core/tagging/api.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py:517]: Yes, I confirm that :+1: 

Captura desde 2024-10-15 19-40-19

Comment (ChrisChV) [reply openedx_tagging/core/tagging/api.py]: Update comment: https://github.com/openedx/openedx-learning/pull/236/commits/639501987f5618a574eeceff6327986869eee2b1

File: openedx_tagging/core/tagging/rest_api/v1/serializers.py
Comment (ChrisChV) [reply openedx_tagging/core/tagging/rest_api/v1/serializers.py]: Updated: https://github.com/openedx/openedx-learning/pull/236/commits/7289b2231ae0f7c0721327552d59501dbd8b2461

**Comment by ChrisChV** [comment]
@bradenmacdonald It's ready for merge.

> Suggested some changes. Thinking ahead, we may also need an API to "convert" the tags if/when the link is broken. That is, if the user links some library into a course, then later breaks the link, the tags that were read-only would need to be changed to is_copied=False so that they can be edited. But we can do that in a separate PR if you prefer.

Sorry I forgot this, yes, we can do this in a separate PR

**Pull request merged**
